service/forward_service.cc

```diff
@@ -136,29 +136,29 @@ class retrying_dispatcher {
 
     future<query::forward_result> dispatch_to_node(netw::msg_addr id, query::forward_request req) {
         if (utils::fb_utilities::is_me(id.addr)) {
-            co_return co_await _forwarder.dispatch_to_shards(req, _tr_info);
+            return _forwarder.dispatch_to_shards(req, _tr_info);
         }
 
         _forwarder._stats.requests_dispatched_to_other_nodes += 1;
 
         // Try to send this forward_request to another node.
-        try {
-            co_return co_await ser::forward_request_rpc_verbs::send_forward_request(
+        return do_with(id, req, [this] (netw::msg_addr& id, query::forward_request& req) -> future<query::forward_result> {
+            return ser::forward_request_rpc_verbs::send_forward_request(
                 &_forwarder._messaging, id, req, _tr_info
-            );
-        } catch(rpc::closed_error& e) {
-            // If a retry has already been done, do a rethrow
-            if (!_retry_available) {
-                flogger.error("failed to send forward_request to node {}: {}", id, e.what());
-                throw;
-            }
-        }
+            ).handle_exception_type([this, &req, &id] (rpc::closed_error& e) -> future<query::forward_result> {
+                if (!_retry_available) {
+                    flogger.error("failed to send forward_request to node {}: {}", id, e.what());
+                    // If a retry has already been done, do a rethrow
+                    return make_exception_future<query::forward_result>(e);
+                }
 
-        // In case of forwarding failure, retry using super-coordinator as a coordinator
-        flogger.warn("retrying forward_request={} on a super-coordinator after failing to send it to {}", req, id);
-        tracing::trace(_tr_state, "retrying forward_request={} on a super-coordinator after failing to send it to", req, id);
-        _retry_available = false;
-        co_return co_await _forwarder.dispatch_to_shards(req, _tr_info);
+                // In case of forwarding failure, retry using super-coordinator as a coordinator
+                flogger.warn("retrying forward_request={} on a super-coordinator after failing to send it to {} ({})", req, id, e.what());
+                tracing::trace(_tr_state, "retrying forward_request={} on a super-coordinator after failing to send it to ({})", req, id, e.what());
+                _retry_available = false;
+                return _forwarder.dispatch_to_shards(req, _tr_info);
+            });
+        });
```

Comment:
After catching an exception caused by sending `forward_request` to a killed node [in this line](
https://github.com/scylladb/scylla/pull/10329/commits/1c611fba83cdcea4fa54b43b95d63bc0dbc403ed#diff-f49e01bb97d1349714f443c2693398f278de3d55234f7c84a8e8af2d50f0b6ddR146), execution was directed inside a [rethrowing if statement](https://github.com/scylladb/scylla/commit/1c611fba83cdcea4fa54b43b95d63bc0dbc403ed#diff-f49e01bb97d1349714f443c2693398f278de3d55234f7c84a8e8af2d50f0b6ddR151). This happened even though `_retry_available` was never set to `false` (If it would be set, a log message would be emitted [in this line](https://github.com/scylladb/scylla/commit/1c611fba83cdcea4fa54b43b95d63bc0dbc403ed#diff-f49e01bb97d1349714f443c2693398f278de3d55234f7c84a8e8af2d50f0b6ddR158)).

Adding a non-trivial destructor (with a logger invocation inside) for `retrying_dispatcher` made that weird behavior disappear.

I've contacted @psarna for a help with finding the reason for unwanted rethrows, but we were unable to find any bugs that could be the cause of it (there was a hope that these fixes would help - https://github.com/scylladb/scylla/pull/10329/commits/e170961b4dda2702f4587650ec71efcc0598c33d but even after their introduction, the problems persisted).